### ISSUE TITLE
datarace: improve detection to all ast.FuncLit and not only ast.GoStmt

### ISSUE THREAD

**Issue by ccoVeille** [mgechev/revive#1411]
**Is your feature request related to a problem? Please describe.**

right now [datarace](https://revive.run/r#datarace) rule is limited to go func() {}

But other can lead to datarace, http.HandlerFunc for example, but in fact anything that reuse an existing variable, in a func literal

Take a look at the example

**Describe the solution you'd like**

Extends the datarace rule to cover has much use case as possible.

Please note I'm aware this might lead to false-positive. But I feel like reusing an error is a bad thing in a func literal.

**Describe alternatives you've considered**

I looked at other linters (out of revive), and datarace seems the best candidate.

**Additional context**

I found it by working on this Capture the flag

https://2024.ctf.link/internal/challenge/fb03748d-7e94-4ca2-8998-a5e0ffcbd761/

Here is a piece of code I worked on to spot all possible issues.

```go
package fixtures

import (
	"net"
	"net/http"
	"os"
)

func main() {
	err := doInit()
	if err != nil {
		panic(err)
	}

	go func() {
		err = securityCheck("goroutine")
		if err != nil { // MATCH /datarace/
			panic(err)
		}

		os.Remove("goroutine_sensitive_file.txt")
	}()

	go processFunc(func() {
		err = securityCheck("processFunc")
		if err != nil { // MATCH /datarace/
			panic(err)
		}
		os.Remove("sensitive_file.txt") // Simulate some sensitive operation
	})

	go repeatFunc(func() {
		err = securityCheck("repeatFunc")
		if err != nil { // MATCH /datarace/
			panic(err)
		}
		os.Remove("goroutine_sensitive_file.txt")
	})

	launchGoroutine(func() {
		err = securityCheck("launchGoroutine")
		if err != nil { // MATCH /datarace/
			panic(err)
		}

		os.Remove("sensitive_file.txt") // Simulate some sensitive operation
	})

	registerFunc(func() {
		err = securityCheck("registerFunc")
		if err != nil { // MATCH /datarace/
			panic(err)
		}
		os.Remove("sensitive_file.txt") // Simulate some sensitive operation
	})

	http.HandleFunc("/delete_handle_func", func(w http.ResponseWriter, r *http.Request) {
		name := r.URL.Query().Get("name")
		if err = securityCheck(name); err != nil { // MATCH /datarace/
			http.Error(w, "security", http.StatusUnauthorized)
			return
		}

		os.Remove(name)
		w.WriteHeader(http.StatusOK)
	})

	http.Handle("/delete_handle", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		name := r.URL.Query().Get("name")
		if err = securityCheck(name); err != nil { // MATCH /datarace/
			http.Error(w, "security", http.StatusUnauthorized)
			return
		}

		os.Remove(name)
		w.WriteHeader(http.StatusOK)
	}))

	listener, err := net.Listen("tcp", ":0")
	if err != nil {
		panic(err)
	}
	http.Serve(listener, nil)
}

// securityCheck is a dummy function that simulates a security check.
func securityCheck(path string) error {
	return nil
}

// doInit is a dummy function that simulates some initialization logic.
func doInit() error {
	return nil
}

func processFunc(f func()) {
	f()
}

func launchGoroutine(f func()) {
	go f()
}

func repeatFunc(f func()) {
	// This function is a placeholder to simulate some repeated execution logic.
	for i := 0; i < 10; i++ {
		f()
	}
}

func registerFunc(f func()) {
	// This function is a placeholder to simulate some registration logic.
	return
}
```
